From 78a13f3fa9a138954a443a393fd5a717def9975a Mon Sep 17 00:00:00 2001
From: Ray Donnelly <mingw.android@gmail.com>
Date: Sat, 27 Sep 2014 15:42:50 +0100
Subject: [PATCH 09/11] MSYS2: openRootTerminal(), getSUCommand(),
 isRootRunning()

MSYS2 behaviour changes for these:
openRootTerminal runs msys2_shell.bat
getSUCommand returns "sh.exe -c "
isRootRunning returns false
---
 src/terminal.cpp  | 7 ++++++-
 src/unixcommand.h | 6 ++++++
 src/wmhelper.cpp  | 5 ++++-
 src/wmhelper.h    | 2 ++
 4 files changed, 18 insertions(+), 2 deletions(-)

diff --git a/src/terminal.cpp b/src/terminal.cpp
index 5085007..0495acc 100644
--- a/src/terminal.cpp
+++ b/src/terminal.cpp
@@ -198,7 +198,12 @@ void Terminal::openRootTerminal()
 {
   if (m_selectedTerminal == ctn_AUTOMATIC)
   {
-    if (UnixCommand::getLinuxDistro() == ectn_MOOOSLINUX && UnixCommand::hasTheExecutable(ctn_RXVT_TERMINAL))
+    if (UnixCommand::getLinuxDistro() == ectn_MSYS2)
+    {
+        QString cmd = ctn_MSYS2_SH + UnixCommand::getMSYS2Root() + QLatin1String("/msys2_shell.bat");
+        m_process->startDetached(cmd);
+    }
+    else if (UnixCommand::getLinuxDistro() == ectn_MOOOSLINUX && UnixCommand::hasTheExecutable(ctn_RXVT_TERMINAL))
     {
       QString cmd = WMHelper::getSUCommand() + " \"" + ctn_RXVT_TERMINAL +
           " -name Urxvt -title Urxvt \"";
diff --git a/src/unixcommand.h b/src/unixcommand.h
index 833716b..81eed8c 100644
--- a/src/unixcommand.h
+++ b/src/unixcommand.h
@@ -112,10 +112,16 @@ public:
   static bool hasTheExecutable( const QString& exeName );
   static bool isAppRunning(const QString &appName, bool justOneInstance = false);
 
+#if defined(_WIN32)
+  static bool isRootRunning(){
+    return false;
+  }
+#else
   static bool isRootRunning(){
     int uid = geteuid();
     return (uid == 0); //Returns TRUE if root is running Octopi
   }
+#endif
 
   static QString getMSYS2Root(){
     if (m_msys2Root.isEmpty())
diff --git a/src/wmhelper.cpp b/src/wmhelper.cpp
index 187087e..4ce7e1a 100644
--- a/src/wmhelper.cpp
+++ b/src/wmhelper.cpp
@@ -292,7 +292,10 @@ QString WMHelper::getGKSUCommand(){
 QString WMHelper::getSUCommand(){
   QString result(ctn_NO_SU_COMMAND);
 
-  if (isXFCERunning() && (UnixCommand::hasTheExecutable(ctn_GKSU_2))){
+  if (UnixCommand::getLinuxDistro() == ectn_MSYS2){
+    return QLatin1String("sh.exe -c ");
+  }
+  else if (isXFCERunning() && (UnixCommand::hasTheExecutable(ctn_GKSU_2))){
     result = getGKSUCommand();
   }
   else if (isKDERunning() && UnixCommand::hasTheExecutable(ctn_KDESU)){
diff --git a/src/wmhelper.h b/src/wmhelper.h
index e96e0b4..ae9405c 100644
--- a/src/wmhelper.h
+++ b/src/wmhelper.h
@@ -25,6 +25,8 @@
 
 const QString ctn_NO_SU_COMMAND("none");
 const QString ctn_ROOT_SH("/bin/sh -c ");
+const QString ctn_MSYS2_SH("cmd.exe /c ");
+
 const QString ctn_KDESU("kdesu");
 const QString ctn_KDE_DESKTOP("kwin");
 const QString ctn_KDE_EDITOR("kwrite");
-- 
2.1.1

